bundles/org.openhab.binding.iammeter/src/main/java/org/openhab/binding/iammeter/internal/IammeterHandler.java

```diff
@@ -0,0 +1,192 @@
+            String content = "";
+            InputStream stream = new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8));
+
+            String response = HttpUtil.executeUrl(httpMethod, url, stream, null, TIMEOUT_MS);
+            JsonElement iammeterDataElement = new JsonParser().parse(response);
+            JsonObject iammeterData = iammeterDataElement.getAsJsonObject();
+            String keyWord = "Data";
+            boolean bRemoveChannels = false;
+            String channelPrefix = "";
+            if (iammeterData.has("data") || (iammeterData.has("Data") && iammeterData.has("SN"))) {
+                bRemoveChannels = true;
+                if (iammeterData.has("data")) {
+                    keyWord = "data";
+                }
+                for (IammeterWEM3080Channel channelConfig : IammeterWEM3080Channel.values()) {
+                    Channel channel = getThing().getChannel(channelConfig.getId());
+                    if (channel != null) {
+                        channelPrefix = IammeterBindingConstants.THING_TYPE_POWERMETER + ":"
+                                + channel.getUID().getThingUID().getId();
+                        State state = getDecimal(
+                                iammeterData.get(keyWord).getAsJsonArray().get(channelConfig.getIndex()).toString(),
+                                channelConfig.getUnit());
+                        updateState(channel.getUID(), state);
+                    }
+                }
+            } else if (iammeterData.has("Datas") && iammeterData.has("SN")) {
+                keyWord = "Datas";
+                for (IammeterWEM3080TChannel channelConfig : IammeterWEM3080TChannel.values()) {
+                    Channel channel = getThing().getChannel(channelConfig.getId());
+                    if (channel != null) {
+                        State state = getDecimal(iammeterData.get(keyWord).getAsJsonArray().get(channelConfig.getRow())
+                                .getAsJsonArray().get(channelConfig.getCol()).toString(), channelConfig.getUnit());
+                        updateState(channel.getUID(), state);
+                    }
+                }
+            }
+            if (bRemoveChannels) {
+                if (!bExtraChannelRemoved) {
+                    thingStructureChanged(channelPrefix);
+                }
+            }
+            stream.close();
+            updateStatus(ThingStatus.ONLINE);
+            // Very rudimentary Exception differentiation
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Communication error with the device: " + e.getMessage());
+        } catch (JsonSyntaxException je) {
+            logger.warn("Invalid JSON when refreshing source {}: {}", getThing().getUID(), je.getMessage());
+        }
+    }
+
+    protected void thingStructureChanged(String channelPrefix) {
```

Comment:
This is not right. You should model the different types as different things. So in the xml you would see 2 things, 1 with all channels and 1 with the limited set of channels. You can then create 2 different classes, or maybe create 2 classes that each parse the specific data. and in the handler factory you can than pass that specific class to this handler based on the specific thing created.

Reply:
thanks for your advice,I have separated those into 2 thingHandler now.